signingworker/worker.py

```diff
@@ -0,0 +1,187 @@
+import random
+import shutil
+import tempfile
+import urlparse
+import arrow
+from jsonschema import ValidationError
+from kombu import Exchange, Queue
+from kombu.mixins import ConsumerMixin
+import redo
+import logging
+import requests
+import sh
+import json
+import taskcluster
+
+from signingworker.task import validate_task
+from signingworker.exceptions import TaskVerificationError, \
+    ChecksumMismatchError, SigningServerError
+from signingworker.utils import my_ip, get_hash
+
+log = logging.getLogger(__name__)
+
+
+class SigningConsumer(ConsumerMixin):
+
+    def __init__(self, connection, exchange, queue_name, worker_type,
+                 taskcluster_config, signing_servers, signing_server_auth,
+                 supported_signing_scopes, tools_checkout):
+        self.connection = connection
+        self.exchange = Exchange(exchange, type='topic', passive=True)
+        self.queue_name = queue_name
+        self.worker_type = worker_type
+        self.routing_key = "*.*.*.*.*.*.{}.#".format(self.worker_type)
+        self.tc_queue = taskcluster.Queue(taskcluster_config)
+        self.signing_servers = signing_servers
+        # TODO: Explicitly set formats in manifest
+        self.signing_formats = ["mar", "gpg"]
+        self.signing_server_auth = signing_server_auth
+        self.supported_signing_scopes = supported_signing_scopes
+        self.tools_checkout = tools_checkout
+        self.cert = os.path.join(self.tools_checkout,
+                                 "release/signing/host.cert")
+
+    def get_consumers(self, consumer_cls, channel):
+        queue = Queue(name=self.queue_name, exchange=self.exchange,
+                      routing_key=self.routing_key, durable=True,
+                      exclusive=False, auto_delete=False)
+        return [consumer_cls(queues=[queue], callbacks=[self.process_message])]
+
+    def process_message(self, body, message):
+        task_id = None
+        run_id = None
+        try:
+            task_id = body["status"]["taskId"]
+            run_id = body["status"]["runs"][-1]["runId"]
+            log.debug("Claiming task %s, run %s", task_id, run_id)
+            self.tc_queue.claimTask(
+                task_id, run_id,
+                {"workerGroup": self.worker_type, "workerId": self.worker_type}
+            )
+            task = self.tc_queue.task(task_id)
+            task_graph_id = task["taskGroupId"]
+            validate_task(task, self.supported_signing_scopes)
+            self.sign(task_id, run_id, task)
+            log.debug("Completing: %s, r: %s", task_id, run_id)
+            self.tc_queue.reportCompleted(task_id, run_id)
+            log.debug("Complete: %s, r: %s, tg: %s", task_id, run_id,
+                      task_graph_id)
+        except taskcluster.exceptions.TaskclusterRestFailure as e:
+            log.exception("TC REST failure, %s", e.status_code)
+            if e.status_code == 409:
+                log.debug("Task already claimed, acking...")
```

Comment:
I checked https://github.com/taskcluster/taskcluster-queue/blob/master/routes/v1.js and it looks like it's safe to ignore tasks with 409